4/m_recaptcha.cpp

```diff
@@ -0,0 +1,228 @@
+         std::string channel_name;
+         while (std::getline(chanstream, channel_name, ','))
+             whitelist_channels.insert(channel_name);
+ 
+         if (whitelist_channels.find(cname) != whitelist_channels.end())
+             return MOD_RES_PASSTHRU;
+ 
+         std::set<int> whitelist_ports;
+         std::stringstream portstream(tag->getString("whitelistports"));
+         std::string port_str;
+         while (std::getline(portstream, port_str, ','))
+             whitelist_ports.insert(std::stoi(port_str));
+ 
+         if (whitelist_ports.find(user->server_sa.port()) != whitelist_ports.end())
+             return MOD_RES_PASSTHRU;
+ 
+         if (captcha_verified.Get(user))
+             return MOD_RES_PASSTHRU;
+ 
+         NotifyUserToVerify(user);
+         return MOD_RES_DENY;
+     }
+ 
+     void ValidateToken(User* user, const std::string& token)
+     {
+         if (!sql)
+         {
+             user->WriteNotice("*** reCAPTCHA: SQL database is not available.");
+             return;
+         }
+ 
+         std::string formatted_query = INSP_FORMAT(
+             "SELECT COUNT(*) FROM recaptcha_app_verificationtoken WHERE token = '{}' AND created_at + INTERVAL '30 minutes' > NOW()",
+             token);
+ 
+         sql->Submit(new ValidateTokenQuery(this, user, captcha_verified), formatted_query);
+     }
+ 
+ private:
+     void NotifyUserToVerify(User* user)
+     {
+         auto& tag = ServerInstance->Config->ConfValue("captchaconfig");
+ 
+         // Generate token
+         std::string token = GenerateToken();
+         std::string hash = GenerateHash(token);
+ 
+         // Debug logs
+         ServerInstance->Logs.Debug(MODNAME, "Generated token: {}", token);
+         ServerInstance->Logs.Debug(MODNAME, "Generated hash: {}", hash);
+ 
+         // Build verification message
+         std::string message = tag->getString("message", "*** reCAPTCHA: Verify your connection at {url}.");
+         std::string link = captcha_url + "?code=" + token + "&hn=" + hash;
+ 
+         size_t pos = message.find("{url}");
+         if (pos != std::string::npos)
+             message.replace(pos, 5, link);
+ 
+         user->WriteNotice(message);
+     }
+ 
+     std::string GenerateToken()
+     {
+         static const char hex_chars[] = "0123456789abcdef";
+         std::string token;
+ 
+         for (int i = 0; i < 64; ++i)
+             token += hex_chars[std::rand() % 16];
+ 
+         return token;
+     }
+ 
+     std::string GenerateHash(const std::string& token)
+     {
+         unsigned char hash[SHA256_DIGEST_LENGTH];
+         SHA256(reinterpret_cast<const unsigned char*>(token.c_str()), token.size(), hash);
+ 
+         std::string hash_string;
+         for (int i = 0; i < SHA256_DIGEST_LENGTH; ++i)
+         {
+             hash_string += "0123456789abcdef"[hash[i] >> 4];
+             hash_string += "0123456789abcdef"[hash[i] & 0xf];
+         }
+ 
+         return hash_string;
+     }
+ };
```

Comment:
You could drop the dependency on OpenSSL here by using the builtin sha256 hash provider (see cloak_sha256 to see how to use this).